### ISSUE TITLE
Fast verification using MPEG7 signatures

### ISSUE THREAD

**Issue by jailuthra** [livepeer/go-livepeer#1991]
## LPMS API

See <https://github.com/livepeer/lpms/pull/253/> for implementation details.

### 1. Generate signature for a segment

Set the boolean `CalcSign = true` in the `TranscodeOptions` struct for each output rendition:

```golang
opts = ffmpeg.TranscodeOptions {
    ...
    Oname: "out.ts",
    CalcSign: true
    ...
}
```

This will create a binary signature file for each output rendition with a suffix `.bin`. Thus `out.ts` will have its signature stored in `out.ts.bin`.

### 2. Compare two signatures for a match

Signatures can be compared either via file paths or via byte buffers:

```golang
func CompareSignatureByBuffer(data1 []byte, data2 []byte) (bool, error)
func CompareSignatureByPath(fname1 string, fname2 string) (bool, error)
```

These functions return `true` if the signatures match, `false` otherwise.

### 1. Configure a stream for verification (via auth-webhook)

Add a`VerificationFreq: uint` field in the `authWebhookResponse` struct:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/mediaserver.go#L104

Parse this field into `core.StreamParamters.VerificationFreq` which the B will use to run verification on roughly `1/VerificationFreq` segments:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/mediaserver.go#L317

### 2. Fetch a pool of Os (via orch-webhook)

Add a `score uint` field in the `webhookResponse` for each O address:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/discovery/wh_discovery.go#L21-L23

For now this `score` is either `0` or `1` for an untrusted or trusted O respectively. Parse the score for each O into the `orchestratorPool` struct:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/discovery/discovery.go#L24-L28

B session manager uses this pool to populate a working set of available Os:
https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/broadcast.go#L315

Note: We also support populating a list of O addresses via a local file passed through a CLI parameter. Ideally we should expect a `score` from both the orch-webhook and local file - but for now we can skip the latter and assume all Os read via the local file are trusted by default.

### 3. Select Os for a verification-enabled stream

Currently the B sends discovery pings to all known Os and selects a single O for transcoding using a stake-weighted mechanism described in https://github.com/livepeer/go-livepeer/issues/1802:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/broadcast.go#L94

For verification-enabled streams select `verificationOrchCount (= 3)` Os after running discovery. Where at least one O is trusted `trustScore == 1` while the rest may be untrusted `trustScore == 0`.

### 4. Send segment to Os for signed transcoding

Submit a segment for transcoding to *all* the Os selected in step 3 if any of the following conditions are met:

i. Current segment is the first segment of a stream i.e. `seg.SeqNo == 0`.
ii. Last transcode attempt timed out or failed, or a refresh occurred and we have a new set of selected Os.
iii. We have probabilistically selected this segment for verification, to meet the frequency constraint `rand.Int() % VerificationFreq == 0`

In step 6 we'll match the signatures generated by all the untrusted Os against the trusted O.

If none of the above conditions are met, submit the segment to the last-used O - as either it is trusted, or it is untrusted but was selected for the last segment, thus it must have succeeded verification.

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/broadcast.go#L598

Signal the O to generate mpeg7 signatures for each of the transcoded renditions - by setting a new field `CalcSign:  true` in the `core.SegTranscodingMetadata` structure:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/segment_rpc.go#L595-L606

This struct is serialized as `net.SegData` and sent via the HTTP request headers:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/segment_rpc.go#L439

### 5. Save MPEG7 signatures to OS

Signal the underlying transcoder to generate mpeg7 signatures by passing the flag set in step 4 `ffmpeg.TranscodeOptions.CalcSign = md.CalcSign` for each of the output renditions:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/core/transcoder.go#L206

Read the signature binary data from the `oname + ".bin"` file into `TranscodedSegmentData`:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/core/transcoder.go#L175

Upload transcoded segments and their signatures to the object store:

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/segment_rpc.go#L173-L174

### 6. Download and verify signatures

If multiple Os were selected in step 4, download the signatures along with the transcoded renditions from each O, pick a rendition at random (suppose 720p) and compare the signature of each untrusted O with the trusted O.

If the signatures match, mark the untrusted O's result as verified so it may be used in step 7. Otherwise mark it as unverified.

~~If a match fails, increment a per-orchestrator failure counter `verificationFailCount`. Temporarily ban an O if its failure counter crosses a threshold `verificationFailThreshold`:~~ (Moved into Step 9)

### 7. Insert verified segments into the playlist

If we selected a single O in step 4 - insert its transcoded segments in the HLS playlist and/or return them via HTTP push API.

Else we selected multiple Os in step 4 - randomly* select one of the Os that got verified in step 6, and use its transcoded renditions.

\* Prefer sticking to the O used for the previous segment, provided it passed verification for current segment.

https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/server/broadcast.go#L818

**Comment by jailuthra** [comment]
missed a task for adding a new orchestrator capability above:

### 8. Add an orchestrator capability for generating mpeg7 signatures

Update the [list of orchestrator capabilities](https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/core/capabilities.go#L44) to include a new `Capability_MPEG7VideoSignature`[*](https://github.com/livepeer/go-livepeer/issues/1993#issuecomment-900648622) capability. The broadcaster will generate [JobCapabilities](https://github.com/livepeer/go-livepeer/blob/67db52b72e8343986fdef0f71c15b8a87a0301ff/core/capabilities.go#L107) to send verification-enabled streams only to Os that signal support for this capability during discovery.

**Comment by jailuthra** [comment]
To keep boundaries between tasks clear, and minimize the scope of the first-pass implementation, we decided to split-out this stretch-goal from step 6, that would be good to have but not a hard requirement for the first-pass implementation of fast verification:

### 9. [Stretch] Ban orchestrators that frequently fail fast verification

After matching the signatures returned by an untrusted O against a trusted O in step 6 - if a match fails, increment a per-orchestrator failure counter `verificationFailCount`. Temporarily ban an O if its failure counter crosses a threshold `verificationFailThreshold`.

**Comment by jailuthra** [comment]
### 5.5 Upload the signature from T -> O in split O/T setups

Currently in split O/T setups only the rendition data is returned in the multipart return -
https://github.com/livepeer/go-livepeer/blob/d1fb9eaad2b0e6d0742879303efd121bd3321574/server/ot_rpc.go#L177

If `md.CalcPerceptualHash` is true, we also need to return the perceptual hash binary data `v.PHash` via the multipart return to the O, which upon receiving it will save it to the object store as usual in step 5.

**Comment by yondonfu** [comment]
Closed by #2086
